code/modules/mob/living/carbon/xenomorph/castes/carrier/carrier.dm

```diff
@@ -25,3 +28,52 @@
 
 	if(statpanel("Game"))
 		stat("Stored Huggers:", "[huggers] / [xeno_caste.huggers_max]")
+
+/mob/living/carbon/xenomorph/carrier/update_icons()
+	. = ..()
+
+	update_hugger_overlays()
+
+/mob/living/carbon/xenomorph/carrier/proc/update_hugger_overlays()
+	if(!hugger_overlays_icon)
+		return
+
+	overlays -= hugger_overlays_icon
+	hugger_overlays_icon.overlays.Cut()
+
+	if(!huggers)
+		hugger_image_index.Cut()
+		return
+
+	update_icon_maths(round(( huggers / xeno_caste.huggers_max ) * 3.999) + 1)
+
+	for(var/i in hugger_image_index)
+		if(stat == DEAD)
+			hugger_overlays_icon.overlays += icon(icon, "clinger_[i] Knocked Down")
+		else if(lying_angle)
+			if((resting || IsSleeping()) && (!IsParalyzed() && !IsUnconscious() && health > 0))
+				hugger_overlays_icon.overlays += icon(icon, "clinger_[i] Sleeping")
+			else
+				hugger_overlays_icon.overlays +=icon(icon, "clinger_[i] Knocked Down")
+		else
+			hugger_overlays_icon.overlays +=icon(icon, "clinger_[i]")
+
+	overlays += hugger_overlays_icon
+
+/mob/living/carbon/xenomorph/carrier/proc/update_icon_maths(number)
+	var/funny_list = list(1,2,3,4)
+	if(length(hugger_image_index) != number)
+		if(length(hugger_image_index) > number)
+			while(length(hugger_image_index) != number)
+				hugger_image_index -= hugger_image_index[length(hugger_image_index)]
+		else
+			while(length(hugger_image_index) != number)
+				for(var/i in hugger_image_index)
+					if(locate(i) in funny_list)
+						funny_list -= i
+				hugger_image_index += funny_list[rand(1,length(funny_list))]
```

Comment:
this proc is overcomplicated and makes no sense to exist when you can just iterate in a for loop for mutable appearance creation

Reply:
yeah, i just generally ported cm code directly, gonna look into it later